### PR TITLE
Trang Frego - API Muncher

### PR DESCRIPTION
**Pull request by tfrego**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I read the API documentation and tried performing queries in Postman.
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | The API Wrapper is an isolated class that is used to grab data from the API. It is in 'lib' because it is not associated with Model, Controller or View. My project would change by creating an API wrapper for each API if more than one API were used.
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | I created a method to list the queried data to be used in the index action and I also created a method to find the specific recipe for the show action. An additional private method was used to create instances of the Recipe class. This allowed me to not rely on the API's json structure when calling the data within my application.
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? |
How does VCR aid in testing an API? | An edge case for the API Wrapper included when there are no entries found for a search criteria. A nominal case is when a user queries for a common recipe and results are returned. VCR aids in testing by recording and replaying the queried action. This aids in reducing the amount of requests I am sending to the API.
What is the Heroku URL of your deployed application? | https://c10-muncher.herokuapp.com/

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Rails.application.routes.draw do
+  root 'recipes#home'
+  get 'recipes/index', to: 'recipes#index', as: 'recipes'
+  get 'recipes/:id', to: 'recipes#show', as: 'recipe'
```
Comment (dHelmgren) [inline config/routes.rb:4]: Could you use `resources` for the index and show pages?

File: test/lib/edamam_api_wrapper_test.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        expect(recipe).must_respond_to :uri
+        expect(recipe).must_respond_to :image
+        expect(recipe).must_respond_to :source
+      end
+    end
+  end
+
+  it 'can find a recipe with a valid id' do
+    VCR.use_cassette('find_recipe') do
+      recipe = EdamamApiWrapper.find_recipe('recipe_7bf4a371c6884d809682a72808da7dc2')
```
Comment (dHelmgren) [inline test/lib/edamam_api_wrapper_test.rb:20]: What if we give it a bad ID? Your code handles it, but a test is good to have!

File: test/controllers/recipes_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    it "can get a list of recipes" do
+      VCR.use_cassette('list_recipes') do
+        get recipes_path, params: { q: "chicken"}
+
+        must_respond_with :ok
+      end
+    end
+
+    it 'will flash warning if search results in 0 entries found' do
```
Comment (dHelmgren) [inline test/controllers/recipes_controller_test.rb:15]: There are some interesting cases you're not covering here:

- What happens if no search term is provided?
- What if the search returns no results?

It might also be worthwhile to add some tests around the paging parameters:

- What happens if they're absent?
- Do you get different results when they change?
- What if you send a bogus value, like a negative page number?

**Comment by dHelmgren** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | Would have liked to see a few more commits
Comprehension questions | Good, missed a question!
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Yes
Semantic HTML | I could cry. This HTML is so good.
Errors are reported to the user | yes
API Wrapper to handle the API requests | yes
Controller testing | Could be better, see comments
Lib testing | see comments
Search Functionality | yes
List Functionality | yes
Show individual item functionality | yes
**Styling** |
List view shows 10 items at a time and/or has pagination | Both!
Attractive and usable UI | Good. Your 10 item per page limit doesn't look great on my machine. It was an either or, so I'd have gone for just pagination.
**API Features** |
The App attributes Edamam | yes
The VCR cassettes do not contain the API key | yes!
**External Resources** |
Link to deployed app on Heroku | yes
**Overall** | Great work overall! You really hit the learning goals here.
